gcc/rust/resolve/rust-forever-stack.hxx

```diff
@@ -0,0 +1,227 @@
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-forever-stack.h"
+#include "rust-rib.h"
+#include "optional.h"
+
+namespace Rust {
+namespace Resolver2_0 {
+
+template <Namespace N>
+bool
+ForeverStack<N>::Node::is_root () const
+{
+  return !parent.has_value ();
+}
+
+template <Namespace N>
+bool
+ForeverStack<N>::Node::is_leaf () const
+{
+  return children.empty ();
+}
+
+template <Namespace N>
+void
+ForeverStack<N>::Node::insert_child (Link link, Node child)
+{
+  auto res = children.insert ({link, child});
+
+  // Do we want to error if the child already exists? Probably not, right?
+  // That's kinda the point, isn't it. So this method always succeeds, right?
+}
+
+// FIXME: Add correct implementation
+template <Namespace N>
+void
+ForeverStack<N>::push (Rib rib, NodeId id, tl::optional<Identifier> path)
+{
+  push_inner (rib, Link (id, path));
+}
+
+template <Namespace N>
+void
+ForeverStack<N>::push_inner (Rib rib, Link link)
+{
+  // If the link does not exist, we create it and emplace a new `Node` with the
+  // current node as its parent. `unordered_map::emplace` returns a pair with
+  // the iterator and a boolean. If the value already exists, the iterator
+  // points to it. Otherwise, it points to the newly emplaced value, so we can
+  // just update our cursor().
+  auto emplace
+    = cursor ().children.emplace (std::make_pair (link, Node (rib, cursor ())));
+
+  auto it = emplace.first;
+  auto didnt_exist = emplace.second;
+
+  rust_debug ("inserting link: Link(%d [%s]): existed? %s", link.id,
+	      link.path.has_value () ? link.path.value ().c_str () : "<anon>",
+	      didnt_exist ? "no" : "yes");
+
+  // We update the cursor
+  cursor () = it->second;
+}
+
+template <Namespace N>
+void
+ForeverStack<N>::pop ()
+{
+  assert (!cursor ().is_root ());
+
+  cursor () = cursor ().parent.value ();
+}
+
+template <Namespace N>
+tl::expected<NodeId, AlreadyInserted>
+ForeverStack<N>::insert (Identifier name, NodeId node)
+{
+  auto &innermost_rib = peek ();
+
+  // So what do we do here - if the Rib has already been pushed in an earlier
+  // pass, we might end up in a situation where it is okay to re-add new names.
+  // Do we just ignore that here? Do we keep track of if the Rib is new or not?
+  // should our cursor have info on the current node like "is it newly pushed"?
```

Comment:
We could encode such information in the type system, mutating a YoungNode into an Grownode, this would prevent state in the cursor as well as state in the node, once grown there would be no way back.

Just throwing random ideas around, don't mind me :laughing: